programs/ziti-edge-tunnel/ziti-edge-tunnel.c

```diff
@@ -229,6 +250,70 @@ static void on_command_resp(const tunnel_result* result, void *ctx) {
     }
 }
 
+void tunnel_enroll_cb(ziti_config *cfg, int status, char *err, void *ctx) {
+    struct add_identity_request_s *add_id_req = ctx;
+
+    tunnel_result result = {
+            .success = false,
+            .error = NULL,
+            .data = NULL,
+    };
+
+    if (status != ZITI_OK) {
+        ZITI_LOG(ERROR, "enrollment failed: %s(%d)", err, status);
+        result.error = "enrollment failed";
+        add_id_req->cmd_cb(&result, add_id_req->cmd_ctx);
+        free(add_id_req);
+        return;
+    }
+
+    FILE *f = add_id_req->add_id_ctx;
+
+    size_t len;
+    char *cfg_json = ziti_config_to_json(cfg, 0, &len);
+
+    if (fwrite(cfg_json, 1, len, f) != len) {
+        ZITI_LOG(ERROR, "failed to write config file");
+        fclose(f);
+        result.error = "failed to write config file";
+        add_id_req->cmd_cb(&result,  add_id_req->cmd_ctx);
+        free(add_id_req);
+        return;
+    }
+
+    free(cfg_json);
+    fflush(f);
+    fclose(f);
+
+    create_or_get_tunnel_identity(add_id_req->identifier, add_id_req->identifier_file_name);
+
+    // send load identity command to the controller
+    tunnel_comand *tnl_cmd = calloc(1, sizeof(tunnel_comand));
+    tnl_cmd->command = TunnelCommand_LoadIdentity;
+    tunnel_load_identity *load_identity_options = calloc(1, sizeof(tunnel_load_identity));
+    load_identity_options->identifier = strdup(add_id_req->identifier);
+    load_identity_options->path = strdup(add_id_req->identifier);
+    size_t json_len;
+    tnl_cmd->data = tunnel_load_identity_to_json(load_identity_options, MODEL_JSON_COMPACT, &json_len);
+    send_tunnel_command(tnl_cmd, add_id_req->cmd_ctx);
+    free_tunnel_load_identity(load_identity_options);
+    free(load_identity_options);
+    free(add_id_req);
+}
+
+static void enroll_ziti_async(uv_async_t *ar) {
+    struct add_identity_request_s *add_id_req = ar->data;
+    uv_loop_t *enroll_loop = ar->loop;
+
+    uv_close((uv_handle_t *) ar, (uv_close_cb) free);
+
+    ziti_enroll_opts enroll_opts = {0};
+    enroll_opts.enroll_name = add_id_req->identifier;
+    enroll_opts.jwt_content = add_id_req->jwt_content;
+
+    ziti_enroll(&enroll_opts, enroll_loop, tunnel_enroll_cb, add_id_req);
```

Comment:
can `add_id_req` be free'd here?

Reply:
add_id_req data is used in the tunnel_enroll_cb callback function. So I am free-ing it in that function.